### PR TITLE
Avoid LOH allocations in Lexer

### PR DESCRIPTION
**Pull request by pierotibou**

### CODE REVIEW

File: Cottle.Test/src/DocumentTester.cs [hunk 1]
```diff
@@ -444,6 +445,43 @@ public void TextTrim (string value, string pattern, string replacement, string e
 			this.AssertRender (value, expected, setting, (s) => {}, (d) => {});
 		}
 
+		[Test]
+		[TestCase("a", 45000, "")]
```
Comment (r3c) [inline Cottle.Test/src/DocumentTester.cs]: CS: please add a whitespace before all opening parenthesis (applies to the whole review).

File: Cottle.Test/src/DocumentTester.cs [hunk 2]
```diff
@@ -444,6 +445,43 @@ public void TextTrim (string value, string pattern, string replacement, string e
...
+			for (int i = 0; i < nbRepeat; i++)
+				sb.Append(inputToRepeat);
+
+			// Add some variables in the text to test a use case where we don't have only text
+			if (!string.IsNullOrEmpty(variable))
+				sb.Append(variable);
+
+			var input = sb.ToString();
+
+			ISetting CottleSettings = new CustomSetting
```
Comment (r3c) [inline Cottle.Test/src/DocumentTester.cs]: No need to change settings here, this should be part of a different test, please stick to defaults.

File: Cottle.Test/src/DocumentTester.cs [hunk 3]
```diff
@@ -444,6 +445,43 @@ public void TextTrim (string value, string pattern, string replacement, string e
 			this.AssertRender (value, expected, setting, (s) => {}, (d) => {});
 		}
 
+		[Test]
+		[TestCase("a", 45000, "")]
+		[TestCase("a", 45000, "(|bob|)")]
+		public void TooLongStringsTest(string inputToRepeat, int nbRepeat, string variable)
```
Comment (r3c) [inline Cottle.Test/src/DocumentTester.cs]: Seems "inputToRepeat" doesn't need to be an input for this test and can be replaced by a constant, as you don't depend on it.

File: Cottle.Test/src/DocumentTester.cs [hunk 4]
```diff
@@ -444,6 +445,43 @@ public void TextTrim (string value, string pattern, string replacement, string e
...
+				sb.Append(variable);
+
+			for (int i = 0; i < nbRepeat; i++)
+				sb.Append(inputToRepeat);
+
+			// Add some variables in the text to test a use case where we don't have only text
+			if (!string.IsNullOrEmpty(variable))
+				sb.Append(variable);
+
+			var input = sb.ToString();
```
Comment (r3c) [inline Cottle.Test/src/DocumentTester.cs]: CS: declare variables at the top of enclosing block.

File: Cottle.Test/src/DocumentTester.cs [hunk 5]
```diff
@@ -444,6 +445,43 @@ public void TextTrim (string value, string pattern, string replacement, string e
...
+		public void TooLongStringsTest(string inputToRepeat, int nbRepeat, string variable)
+		{
+			var sb = new StringBuilder();
+
+			// Add some variables in the text to test a use case where we don't have only text
+			if (!string.IsNullOrEmpty(variable))
+				sb.Append(variable);
+
+			for (int i = 0; i < nbRepeat; i++)
+				sb.Append(inputToRepeat);
```
Comment (r3c) [inline Cottle.Test/src/DocumentTester.cs]: Can be replaced by `sb.Append(inputToRepeat, nbRepeat)` (or just `sb.Append('a', nbRepeat)`).

File: Cottle/src/Parsers/Default/Lexer.cs [hunk 6]
```diff
@@ -38,6 +38,8 @@ public int Line
 		#endregion
 
 		#region Attributes
+		// A buffer containing more than 85000 bytes will be allocated on LOH
+		private const int MaxBufferSize = 84000 / UnicodeEncoding.CharSize;
```
Comment (r3c) [inline Cottle/src/Parsers/Default/Lexer.cs]: Replace `UnicodeEncoding.CharSize` with `sizeof (char)` for an encoding-agnostic expression (and incidentally fix the DotNetCore compile error).

File: Cottle/src/Parsers/Default/Lexer.cs [hunk 7]
```diff
@@ -379,22 +381,29 @@ private Lexem NextRaw ()
 
 						this.cursors[candidate] = next;
 
-						// No lexem matched for this cursor, continue loop
-						if (next.State == null || next.State.Type == LexemType.None)
+						// No lexem matched for this cursor (and no risk of LOH allocation), continue loop
+						if ((next.State == null || next.State.Type == LexemType.None) && buffer.Length < MaxBufferSize)
```
Comment (r3c) [inline Cottle/src/Parsers/Default/Lexer.cs]: You can't break from this loop until you found a candidate or are sure no one match. Current code will break if some valid lexem exceeds your LOH length (which is extremely unlikely, yet acceptable). There is probably a missing test case here.

**Comment by pierotibou** [comment]
And you told Guillaume on a different pull request that he shouldn't work
on week-ends... ;)

**Comment by r3c** [comment]
That's not me, just some automated comments :)

File: Cottle.Test/src/DocumentTester.cs [hunk 8]
```diff
@@ -1,9 +1,11 @@
-﻿using System;
+﻿System;
```
Comment (r3c) [inline Cottle.Test/src/DocumentTester.cs]: This change is probably not wanted, and seems to be breaking the build.

**Comment by r3c** [comment]
Probably not needed anymore, I'll close this PR soon unless you think some pending changes are not available in master yet.

**Pull request closed without merging**